File: src/main/java/nl/dannyjelsma/soundboardbot/controllers/SoundController.java
package nl.dannyjelsma.soundboardbot.controllers;

import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import nl.dannyjelsma.soundboardbot.APIKeys;
import nl.dannyjelsma.soundboardbot.SoundBoardBot;
import nl.dannyjelsma.soundboardbot.audio.Sound;
import nl.dannyjelsma.soundboardbot.audio.SoundManager;
import nl.dannyjelsma.soundboardbot.audio.SoundScheduler;
import nl.dannyjelsma.soundboardbot.exceptions.InvalidURLException;
import nl.dannyjelsma.soundboardbot.exceptions.SoundAlreadyExistsException;
import nl.dannyjelsma.soundboardbot.exceptions.SoundNotFoundException;
import nl.dannyjelsma.soundboardbot.exceptions.UnauthorizedException;
import nl.dannyjelsma.soundboardbot.tts.TTSClient;
import org.springframework.web.bind.annotation.*;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
public class SoundController {

    @GetMapping("/soundboard/sounds")
    public List<Sound> getSounds(@RequestHeader("X-API-KEY") String apiKey) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        return SoundManager.getInstance().getSounds();
    }

    @GetMapping("/soundboard/testkey")
    public boolean testAPIKey(@RequestHeader("X-API-KEY") String apiKey) {
        return isKeyValid(apiKey);
    }

    @DeleteMapping("/soundboard/sound/{soundName}")
    public void deleteSound(@RequestHeader("X-API-KEY") String apiKey, @PathVariable String soundName) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        File file = new File("." + File.separator + "sounds" + File.separator + soundName);
        if (!file.exists()) {
            throw new SoundNotFoundException();
        } else {
            File[] files = file.listFiles();

            if (files != null && files.length > 0) {
                for (File dirFile : files) {
                    dirFile.delete();
                }
            }

            SoundManager manager = SoundManager.getInstance();
            manager.clearSounds();
            manager.loadSounds();
            file.delete();
        }
    }

    @PostMapping("/soundboard/sound/{soundName}")
    public void uploadSound(@RequestHeader("X-API-KEY") String apiKey, @PathVariable String soundName, @RequestParam(value = "url") String url) throws InterruptedException, IOException {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        try {
            new URL(url);
        } catch (MalformedURLException e) {
            throw new InvalidURLException();
        }

        File file = new File("." + File.separator + "sounds" + File.separator + soundName + File.separator + soundName + ".m4a");

        if (file.exists()) {
            throw new SoundAlreadyExistsException();
        } else {
            file.getParentFile().mkdir();
        }

        Process process = new ProcessBuilder("." + File.separator + "yt-dlp", "-f", "bestaudio[ext=m4a]", "--extract-audio", "--audio-format", "mp3", "-o", file.getAbsolutePath(), url.toString()).redirectErrorStream(true).start();
        BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream()));
        String line;
        while ((line = in.readLine()) != null) {
            System.out.println(line);
        }
        process.waitFor();
        file.delete();

        SoundManager manager = SoundManager.getInstance();
        manager.clearSounds();
        manager.loadSounds();
    }


    @GetMapping("/soundboard/sound/{soundName}")
    public Sound getSound(@RequestHeader("X-API-KEY") String apiKey, @PathVariable String soundName) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        File file = new File("." + File.separator + "sounds" + File.separator + soundName);
        if (!file.exists()) {
            throw new SoundNotFoundException();
        }

        return SoundManager.getInstance().getSound(soundName);
    }

    @GetMapping("/soundboard/state")
    public Map<String, Object> getState(@RequestHeader("X-API-KEY") String apiKey) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        Map<String, Object> state = new HashMap<>();
        SoundManager manager = SoundManager.getInstance();
        SoundScheduler scheduler = manager.getScheduler();

        List<AudioTrack> musicQueue = scheduler.getMusicQueue();
        List<String> musicNames = new ArrayList<>();
        List<AudioTrack> soundboardQueue = scheduler.getMusicQueue();
        List<String> soundboardNames = new ArrayList<>();
        AudioTrack playingTrack = manager.getAudioPlayer().getPlayingTrack();
        String playingTrackName;

        if (playingTrack != null) {
            playingTrackName = playingTrack.getInfo().title + " - " + playingTrack.getInfo().author;
        } else {
            playingTrackName = "None";
        }

        int volume = manager.getAudioPlayer().getVolume();

        for (AudioTrack track : musicQueue) {
            musicNames.add(track.getInfo().title + " - " + track.getInfo().author);
        }

        for (AudioTrack track : soundboardQueue) {
            soundboardNames.add(track.getInfo().title + " - " + track.getInfo().author);
        }

        state.put("musicQueue", musicNames);
        state.put("soundboardQueue", soundboardNames);
        state.put("playingTrack", playingTrackName);
        state.put("volume", volume);

        return state;
    }

    @GetMapping("/soundboard/skip")
    public void skipSound(@RequestHeader("X-API-KEY") String apiKey) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        SoundManager manager = SoundManager.getInstance();
        manager.getAudioPlayer().stopTrack();
    }

    @GetMapping("/soundboard/stop")
    public void stopSound(@RequestHeader("X-API-KEY") String apiKey) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        SoundManager manager = SoundManager.getInstance();
        manager.getScheduler().stop();
    }

    @PostMapping("/soundboard/TTS")
    public void playTTS(@RequestHeader("X-API-KEY") String apiKey, @RequestParam(value = "voice") String voice, @RequestParam(value = "message") String message) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        TTSClient tts = new TTSClient(voice);
        String url = tts.getSpeechURL(message);

        tts.playSpeechFromURL(url);
    }

    @GetMapping("/soundboard/volume")
    public int getVolume(@RequestHeader("X-API-KEY") String apiKey) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        SoundManager manager = SoundManager.getInstance();
        return manager.getAudioPlayer().getVolume();
    }

    @PostMapping("/soundboard/volume")
    public void setVolume(@RequestHeader("X-API-KEY") String apiKey, @RequestParam(value = "volume") int volume) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        SoundManager manager = SoundManager.getInstance();
        manager.getAudioPlayer().setVolume(volume);
    }

    @GetMapping("/soundboard/play/{soundName}")
    public void playSound(@RequestHeader("X-API-KEY") String apiKey, @PathVariable String soundName) {
        if (!isKeyValid(apiKey)) {
            throw new UnauthorizedException();
        }

        SoundManager manager = SoundManager.getInstance();
        if (!manager.hasSound(soundName)) {
            throw new SoundNotFoundException();
        } else {
            Sound sound = manager.getSound(soundName);

            manager.getAudioPlayerManager().loadItemOrdered(manager, sound.file().getAbsolutePath(), new AudioLoadResultHandler() {
                @Override
                public void trackLoaded(AudioTrack track) {
                    manager.getScheduler().queue(track, true);
                }

                @Override
                public void playlistLoaded(AudioPlaylist playlist) {
                    for (AudioTrack track : playlist.getTracks()) {
                        manager.getScheduler().queue(track, true);
                    }
                }

                @Override
                public void noMatches() {
                    throw new SoundNotFoundException();
                }

                @Override
                public void loadFailed(FriendlyException exception) {
                    throw exception;
                }
            });
        }
    }

    private boolean isKeyValid(String key) {
        APIKeys keys = SoundBoardBot.getApiKeys();

        return keys.containsKey(key);
    }
}

File: src/main/java/nl/dannyjelsma/soundboardbot/controllers/SoundController.kt
package nl.dannyjelsma.soundboardbot.controllers

import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import nl.dannyjelsma.soundboardbot.SoundBoardBot
import nl.dannyjelsma.soundboardbot.audio.Sound
import nl.dannyjelsma.soundboardbot.audio.SoundManager
import nl.dannyjelsma.soundboardbot.exceptions.InvalidURLException
import nl.dannyjelsma.soundboardbot.exceptions.SoundAlreadyExistsException
import nl.dannyjelsma.soundboardbot.exceptions.SoundNotFoundException
import nl.dannyjelsma.soundboardbot.exceptions.UnauthorizedException
import nl.dannyjelsma.soundboardbot.tts.TTSClient
import org.springframework.util.MultiValueMap
import org.springframework.web.bind.annotation.*
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.net.URL
import java.util.*

@RestController
class SoundController {
    @GetMapping("/soundboard/sounds")
    fun getSounds(@RequestHeader("X-API-KEY") apiKey: String): List<Sound> {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        return SoundManager.getSounds()
    }

    @GetMapping("/soundboard/testkey")
    fun testAPIKey(@RequestHeader("X-API-KEY") apiKey: String): Boolean {
        return isKeyValid(apiKey)
    }

    @DeleteMapping("/soundboard/sound/{soundName}")
    fun deleteSound(@RequestHeader("X-API-KEY") apiKey: String, @PathVariable soundName: String) {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        val file = File("." + File.separator + "sounds" + File.separator + soundName)
        if (!file.exists()) {
            throw SoundNotFoundException()
        } else {
            val files = file.listFiles()
            if (files != null && files.isNotEmpty()) {
                for (dirFile in files) {
                    dirFile.delete()
                }
            }

            SoundManager.clearSounds()
            SoundManager.loadSounds()
            file.delete()
        }
    }

    @PostMapping("/soundboard/sound/{soundName}")
    @Throws(InterruptedException::class, IOException::class)
    fun uploadSound(@RequestHeader("X-API-KEY") apiKey: String,
                    @PathVariable soundName: String,
                    @RequestBody body: MultiValueMap<String, String>) {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        try {
            URL(Objects.requireNonNull(body.getFirst("url")))
        } catch (e: Exception) {
            throw InvalidURLException()
        }

        val file =
            File("." + File.separator + "sounds" + File.separator + soundName + File.separator + soundName + ".m4a")
        if (file.exists()) {
            throw SoundAlreadyExistsException()
        } else {
            file.parentFile.mkdir()
        }

        val process = ProcessBuilder("." + File.separator + "yt-dlp",
            "-f",
            "bestaudio[ext=m4a]",
            "--extract-audio",
            "--audio-format",
            "mp3",
            "-o",
            file.absolutePath,
            body.toString()).redirectErrorStream(true).start()

        val `in` = BufferedReader(InputStreamReader(process.inputStream))
        var line: String?
        while (`in`.readLine().also { line = it } != null) {
            println(line)
        }

        process.waitFor()
        file.delete()
        SoundManager.clearSounds()
        SoundManager.loadSounds()
    }

    @GetMapping("/soundboard/sound/{soundName}")
    fun getSound(@RequestHeader("X-API-KEY") apiKey: String, @PathVariable soundName: String): Sound? {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        val file = File("." + File.separator + "sounds" + File.separator + soundName)
        if (!file.exists()) {
            throw SoundNotFoundException()
        }

        return SoundManager.getSound(soundName)
    }

    @GetMapping("/soundboard/state")
    fun getState(@RequestHeader("X-API-KEY") apiKey: String): Map<String, Any> {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        val state: MutableMap<String, Any> = HashMap()
        val musicQueue = SoundManager.scheduler.musicQueue
        val musicNames: MutableList<String> = ArrayList()
        val soundboardQueue = SoundManager.scheduler.musicQueue
        val soundboardNames: MutableList<String> = ArrayList()
        val playingTrack = SoundManager.audioPlayer.playingTrack
        val playingTrackName: String = if (playingTrack != null) {
            playingTrack.info.title + " - " + playingTrack.info.author
        } else {
            "None"
        }

        val volume = SoundManager.audioPlayer.volume
        for (track in musicQueue) {
            musicNames.add(track.info.title + " - " + track.info.author)
        }

        for (track in soundboardQueue) {
            soundboardNames.add(track.info.title + " - " + track.info.author)
        }

        state["musicQueue"] = musicNames
        state["soundboardQueue"] = soundboardNames
        state["playingTrack"] = playingTrackName
        state["volume"] = volume
        return state
    }

    @GetMapping("/soundboard/skip")
    fun skipSound(@RequestHeader("X-API-KEY") apiKey: String) {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        SoundManager.audioPlayer.stopTrack()
    }

    @GetMapping("/soundboard/stop")
    fun stopSound(@RequestHeader("X-API-KEY") apiKey: String) {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        SoundManager.scheduler.stop()
    }

    @PostMapping("/soundboard/TTS")
    fun playTTS(@RequestHeader("X-API-KEY") apiKey: String, @RequestBody body: MultiValueMap<String, String>) {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        val voice: String? = body.getFirst("voice")
        val message: String? = body.getFirst("message")

        if (voice != null && message != null) {
            val tts = TTSClient(voice)
            val url = tts.getSpeechURL(message)

            if (url != null) {
                tts.playSpeechFromURL(url)
            }
        }
    }

    @GetMapping("/soundboard/volume")
    fun getVolume(@RequestHeader("X-API-KEY") apiKey: String): Int {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        return SoundManager.audioPlayer.volume
    }

    @PostMapping("/soundboard/volume")
    fun setVolume(@RequestHeader("X-API-KEY") apiKey: String, @RequestBody body: MultiValueMap<String, Int>) {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        val volume = body.getFirst("volume")

        if (volume != null) {
            SoundManager.audioPlayer.volume = volume
        }
    }

    @GetMapping("/soundboard/play/{soundName}")
    fun playSound(@RequestHeader("X-API-KEY") apiKey: String, @PathVariable soundName: String) {
        if (!isKeyValid(apiKey)) {
            throw UnauthorizedException()
        }

        if (!SoundManager.hasSound(soundName)) {
            throw SoundNotFoundException()
        } else {
            val sound = SoundManager.getSound(soundName)
            if (sound != null) {
                SoundManager.audioPlayerManager.loadItemOrdered(SoundManager, sound.file.absolutePath,
                    object : AudioLoadResultHandler {
                        override fun trackLoaded(track: AudioTrack) {
                            SoundManager.scheduler.queue(track, true)
                        }

                        override fun playlistLoaded(playlist: AudioPlaylist) {
                            for (track in playlist.tracks) {
                                SoundManager.scheduler.queue(track, true)
                            }
                        }

                        override fun noMatches() {
                            throw SoundNotFoundException()
                        }

                        override fun loadFailed(exception: FriendlyException) {
                            throw exception
                        }
                    })
            }
        }
    }

    private fun isKeyValid(key: String): Boolean {
        return SoundBoardBot.apiKeys.containsKey(key)
    }
}